File: trunk/nao/head_orienter/src/head_orienter.cpp
#include <ros/ros.h>
#include <nao_msgs/HeadAngles.h>
#include <geometry_msgs/PoseStamped.h>
#include <tf/transform_datatypes.h>

ros::Publisher* pubPtr = NULL;

void callback(const geometry_msgs::PoseStampedConstPtr& msg) {

  tf::Quaternion q;
  tf::quaternionMsgToTF(msg->pose.orientation, q);

  double y, p, r;
  btMatrix3x3 R;
  R.setRotation(q);
  R.getEulerYPR (y, p, r);

  nao_msgs::HeadAngles headAngleMsg;
  headAngleMsg.yaw = r; // swapped, to bring android phone frame into coincidence with robot head frame
  headAngleMsg.pitch = p;
  headAngleMsg.absolute = 1; // 1 means "absolute" position

  if(pubPtr != NULL) pubPtr->publish(headAngleMsg);

}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "head_orienter");
  ros::NodeHandle nh;

  ros::Publisher pub = nh.advertise<nao_msgs::HeadAngles>("/head_angles", 1);
  pubPtr = &pub;

  ros::Subscriber sub = nh.subscribe("/android/orientation", 1, callback);

  ros::Rate r(10); // limit rate to 10 Hz

  while(ros::ok()) {
    ros::spinOnce();
    r.sleep();
  }

  return 0;

}

File: nao/head_orienter/src/HeadOrienter.py
#!/usr/bin/env python

import roslib; roslib.load_manifest('head_orienter')
import rospy
import tf
from nao_msgs.msg import HeadAngles
from geometry_msgs.msg import PoseStamped

class HeadOrienterNode:

    def __init__(self):
        self.headPub = rospy.Publisher('/head_angles', HeadAngles)
        self.orientationSub = rospy.Subscriber('/android/orientation',PoseStamped,self.orientationCallback)
        rospy.init_node('phone_head')

    def orientationCallback(self, msg):
        #print msg
        q = msg.pose.orientation
        [r,p,y] = tf.transformations.euler_from_quaternion([q.x,q.y,q.z,q.w])

        # 1 means "absolute" position
        #self.headPub.publish(HeadAngles(y,p,1))
        self.headPub.publish(HeadAngles(r,p,1)) # swapped, to bring android phone frame into coincidence with robot head frame

	# FIXME: filter noisy IMU (mean/median filter?, disregard huge spikes, add deadzone, quantize)

if __name__ == '__main__':
    node = HeadOrienterNode()
    rospy.spin()


